torchvision/prototype/models/vision_transformer.py

```diff
@@ -230,6 +231,69 @@ def forward(self, x: torch.Tensor):
 
         return x
 
+    def from_checkpoint(self, model_state: Dict, strict: bool = True, reset_heads=False):
+        # Shape of pos_embedding is (1, seq_length, hidden_dim)
+        pos_embedding = model_state["encoder.pos_embedding"]
+        n, seq_length, hidden_dim = pos_embedding.shape
+        if n != 1:
+            raise ValueError(f"Unexpected position embedding shape: {pos_embedding.shape}")
+        if hidden_dim != self.hidden_dim:
+            raise ValueError(f"Position embedding hidden_dim incorrect: {hidden_dim}" f", expected: {self.hidden_dim}")
+        new_seq_length = self.seq_length
+
+        # Need to interpolate the weights for the position embedding.
+        # We do this by reshaping the positions embeddings to a 2d grid, performing
+        # an interpolation in the (h, w) space and then reshaping back to a 1d grid.
+        if new_seq_length != seq_length:
+            # The class token embedding shouldn't be interpolated so we split it up.
+            seq_length -= 1
+            new_seq_length -= 1
+            pos_embedding_token = pos_embedding[:, :1, :]
+            pos_embedding_img = pos_embedding[:, 1:, :]
+
+            # (1, seq_length, hidden_dim) -> (1, hidden_dim, seq_length)
+            pos_embedding_img = pos_embedding_img.permute(0, 2, 1)
+            seq_length_1d = int(math.sqrt(seq_length))
+            torch._assert(seq_length_1d * seq_length_1d == seq_length, "seq_length is not a perfect square!")
+
+            logging.info(
+                "Interpolating the position embeddings from image "
+                f"{seq_length_1d * self.patch_size} to size {self.image_size}"
+            )
+
+            # (1, hidden_dim, seq_length) -> (1, hidden_dim, seq_l_1d, seq_l_1d)
+            pos_embedding_img = pos_embedding_img.reshape(1, hidden_dim, seq_length_1d, seq_length_1d)
+            new_seq_length_1d = self.image_size // self.patch_size
```

Comment:
Is it possible that one might be interested in providing another value for `new_seq_length_1d` which is estimated differently? I wonder if it would make sense to make this value an argument to the method and eliminate the need for accessing `self.image_size` and `self.patch_size` here. This will also allow you to remove this method from the nn.Module and instead offer it as a standalone utility method.

One way you could potentially do this is by offering the following signature:
```python
def interpolate_embeddings(model_state: Dict, new_seq_length: int, new_seq_length_1d: int, reset_heads=False) -> Dict
```

If the `new_seq_length_1d` is always calculated as above, then an alternative signature could be:
```python
def interpolate_embeddings(model_state: Dict, new_seq_length: int, image_size: int, patch_size: int, reset_heads=False) -> Dict
```

Thoughts?

Reply:
Thanks for your comment.

Actually, if we decide to change the API like this, then we assume the users know exactly how to compute  `seq_length_1d` and `seq_length`.
i.e. they should know `seq_length_1d = image_size // patch_size` & `seq_length = (image_size // patch_size) ** 2 + 1`

Do you feel this is a reasonable assumption?
cc @mannatsingh @datumbox

Reply:
If the above is the only reasonable way to set the `seq_length*` variables then we can pass `image_size` and `patch_size` and make the estimation internally. If we anticipate that there will be more reasonable ways to set the vars, then we could pass `seq_length` and `new_seq_length`.

I don't have a strong opinion over how to parameterize this. I just think this would be better served in a utility method than in a method which is attached to the model and explicitly does the loading of the dictionary.

Reply:
Sure - we can do that. Where shall we park this utility function? In `vision_transformer.py` or in `prototype/utils`?

Reply:
> If we anticipate that there will be more reasonable ways to set the vars, then we could pass seq_length and new_seq_length.

I don't know whether there are use cases like this, probably @fmassa @mannatsingh  know better here?

Reply:
> Sure - we can do that. Where shall we park this utility function? In vision_transformer.py or in prototype/utils?

I think we should leave it on vision_transformer.py :)

Reply:
I think the only way the seq length is calculated is based on image size and patch size, so there shouldn't be a need to take `seq_length` as a parameter directly. In fact, it will probably cause a bit of confusion.

Reply:
Sounds good. Let's make `image_size` and `patch_size` parameters of the method and estimate the `seq_length_1d` and `seq_length` within the utility as @sallysyw suggested. :)